components/hab/src/cli_v4/pkg/bulk_upload.rs

```diff
@@ -0,0 +1,85 @@
+// Implemenatation of `hab pkg bulkupload`
+
+use clap_v4 as clap;
+
+use std::path::PathBuf;
+
+use clap::{ArgAction,
+           Parser};
+
+use habitat_common::{cli::clap_validators::DirExistsValueParser,
+                     ui::UI};
+
+use habitat_core::{crypto::keys::KeyCache,
+                   ChannelIdent};
+
+use habitat_api_client::BuildOnUpload;
+
+use crate::{cli_v4::utils::{AuthToken,
+                            BldrUrl},
+            command::pkg::bulkupload,
+            error::Result as HabResult};
+
+#[derive(Debug, Clone, Parser)]
+#[command(arg_required_else_help = true,
+          help_template = "{name} {version} {author-section} {about-section} \n{usage-heading} \
+                           {usage}\n\n{all-args}\n")]
+pub(crate) struct PkgBulkUploadOptions {
+    #[command(flatten)]
+    bldr_url: BldrUrl,
+
+    #[command(flatten)]
+    auth_token: AuthToken,
+
+    /// Optional additional release channel to upload package to. Packages are always uploaded
+    /// to `unstable`, regardless of the value of this option
+    #[arg(name = "CHANNEL", short = 'c', long = "channel")]
+    channel: Option<ChannelIdent>,
+
+    /// Skip checking availability of package and force uploads, potentially overwriting a
+    /// stored copy of a package
+    #[arg(name = "FORCE", long = "force", action = ArgAction::SetTrue)]
+    force: bool,
+
+    // TODO: This option is to be removed?
+    /// Enable auto-build for all packages in this upload. Only applicable to SaaS Builder
+    #[arg(name = "AUTO_BUILD", long = "auto-build", action = ArgAction::SetTrue)]
+    auto_build: bool,
+
+    /// Skip the confirmation prompt and automatically create origins that do not exist in the
+    /// target Builder
+    #[arg(name = "AUTO_CREATE_ORIGINS", long = "auto-create-origins", action = ArgAction::SetTrue)]
+    auto_create_channels: bool,
+
+    // TODO: Add Path Exists validator.
+    /// Directory Path from which artifacts will be uploaded
+    #[arg(name = "UPLOAD_DIRECTORY", value_parser = DirExistsValueParser)]
+    upload_directory: PathBuf,
```

Comment:
The value parser is choking here on a valid directory:

```
wrock@PF4YL8GK:/mnt/c/dev/habitat$ /hab/pkgs/mwrock/hab/2.0.12/20250108201920/bin/hab pkg bulkupload dev-docs
thread 'main' panicked at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.5.9/src/parser/error.rs:32:9:
Mismatch between definition and access of `UPLOAD_DIRECTORY`. Could not downcast to TypeId { t: (16882638169297981978, 4285806014341565508) }, need to downcast to TypeId { t: (3406769651295146987, 5568765126405537519) }

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I think it is because the parser is returning a string but this expects a PathBuf.